File: Behavior/Interpreter/TypeScript/interpreter.ts
// 表达式接口
interface Expression {
    interpret(context: InterpreterContext): number;
}

// 数字表达式
class NumberExpression implements Expression {
    private value: number;

    constructor(value: string) {
        this.value = parseInt(value);
    }

    interpret(context: InterpreterContext): number {
        return this.value;
    }
}

// 加法表达式
class AddExpression implements Expression {
    private leftExpression: Expression;
    private rightExpression: Expression;

    constructor(left: Expression, right: Expression) {
        this.leftExpression = left;
        this.rightExpression = right;
    }

    interpret(context: InterpreterContext): number {
        return this.leftExpression.interpret(context) + this.rightExpression.interpret(context);
    }
}

// 减法表达式
class SubtractExpression implements Expression {
    private leftExpression: Expression;
    private rightExpression: Expression;

    constructor(left: Expression, right: Expression) {
        this.leftExpression = left;
        this.rightExpression = right;
    }

    interpret(context: InterpreterContext): number {
        return this.leftExpression.interpret(context) - this.rightExpression.interpret(context);
    }
}

// 变量表达式
class VariableExpression implements Expression {
    private name: string;

    constructor(name: string) {
        this.name = name;
    }

    interpret(context: InterpreterContext): number {
        return context.lookup(this.name);
    }

    getName(): string {
        return this.name;
    }
}

// 解释器上下文
class InterpreterContext {
    private variables: Map<string, number>;

    constructor() {
        this.variables = new Map<string, number>();
    }

    assign(variable: VariableExpression, value: number) {
        this.variables.set(variable.getName(), value);
    }

    lookup(name: string): number {
        if (!this.variables.has(name)) {
            throw new Error(`Undefined variable ${name}`);
        }
        return this.variables.get(name)!;
    }
}

// 客户端代码
function main() {
    const context = new InterpreterContext();

    // 定义一些变量
    const x = new VariableExpression("x");
    const y = new VariableExpression("y");

    // 为变量赋值
    context.assign(x, 10);
    context.assign(y, 20);

    // 构造一个简单的表达式：x + y - 5
    const expression = new SubtractExpression(
        new AddExpression(x, y),
        new NumberExpression("5")
    );

    // 解释表达式
    const result = expression.interpret(context);

    console.log(result); // 输出结果应为25
}

main();

File: Behavior/Interpreter/Cpp/interpreter.cpp
#include <iostream>

class Expression
{
public:
    virtual int interpret() = 0;
    virtual ~Expression() {}
};

class NumberExpression : public Expression
{
    int number;

public:
    NumberExpression(int num) : number(num) {}

    int interpret() override
    {
        return number;
    }
};

class AddExpression : public Expression
{
    Expression *left;
    Expression *right;

public:
    AddExpression(Expression *l, Expression *r) : left(l), right(r) {}

    int interpret() override
    {
        return left->interpret() + right->interpret();
    }

    ~AddExpression()
    {
        delete left;
        delete right;
    }
};

class SubtractExpression : public Expression
{
    Expression *left;
    Expression *right;

public:
    SubtractExpression(Expression *l, Expression *r) : left(l), right(r) {}

    int interpret() override
    {
        return left->interpret() - right->interpret();
    }

    ~SubtractExpression()
    {
        delete left;
        delete right;
    }
};

int main()
{
    Expression *left = new NumberExpression(1);
    Expression *right = new NumberExpression(2);
    Expression *addition = new AddExpression(left, right);

    std::cout << "1 + 2 = " << addition->interpret() << std::endl;

    delete addition; // 会自动删除左右表达式

    return 0;
}
